File: ruby/035_PreserveConnectivity.rb
# 035 - Preserve Connectivity（★7）
# 木DP, 最近共通祖先(LCA), 座標圧縮

N = gets.to_i
BIT = N.bit_length
edges = Array.new(N + 1) { [] }
(N - 1).times do
  a, b = gets.split.map(&:to_i)
  edges[a] << b
  edges[b] << a
end

# 親の位置を記録する配列
parents = Array.new(BIT) { [0] }
# 頂点0からの深さ
depth = [0]
# dfs順の番号をメモ
orders = [0]
# 現在の順番
cur_ord = 1
dfs = -> (pos = 1, pre = 0) do
  parents[0][pos] = pre
  depth[pos] = depth[pre] + 1
  orders[pos] = cur_ord
  cur_ord += 1
  edges[pos].each do |nxt|
    next if nxt == pre
    dfs.(nxt, pos)
  end
end

dfs.()

# ダブリング
(BIT - 1).times do |i|
  1.upto(N) do |j|
    parents[i + 1][j] = parents[i][parents[i][j]]
  end
end

# LCA(最近共通祖先)
lca = -> (u, v) do
  u, v = v, u if depth[u] < depth[v]
  diff = depth[u] - depth[v]

  # u と v の深さを合わせる
  BIT.times do |i|
    next if diff[i].zero?
    u = parents[i][u]
  end

  return u if u == v

  (BIT - 1).downto(0) do |i|
    # 2のi乗個上の頂点ではなく1個上の頂点が知るため
    next if parents[i][u] == parents[i][v]
    u = parents[i][u]
    v = parents[i][v]
  end
  parents[0][u]
end

ans = gets.to_i.times.map do |i|
  k, *v = gets.split.map(&:to_i)
  # 巡回順に並び替え
  v.sort_by! { orders[_1] }

  k.times.reduce(0) do |acc, j|
    # v[j]までの深さ と v[(j + 1) % k]との最近共通祖先までの深さの差が必要な辺の数
    acc + depth[v[j]] - depth[lca.(v[j], v[(j + 1) % k])]
  end
end

puts ans

File: javascript/035_PreserveConnectivity.js
// 035 - Preserve Connectivity（★7）
// オイラーツアー
// 最近共通祖先(LCA)
// 座標圧縮

const BIT = 21;

const input = require("fs").readFileSync("/dev/stdin", "utf8").split("\n");

let n = 0;
const N = Number(input[n++]);
const G = [...Array(N)].map(() => []);
for (let i = 0; i < N - 1; i++) {
  const [a, b] = input[n++].split(" ").map(j => Number(j) - 1);
  G[a].push(b);
  G[b].push(a);
}

const parents = [...Array(BIT)].map(() => Array(N).fill(0));
const depth = Array(N).fill(-1);
depth[0] = 0;
const orders = Array(N).fill(-1);
let current_order = 0;

const dfs = (pos = 0, pre = 0) => {
  parents[0][pos] = pre;
  orders[pos] = current_order++;
  G[pos].forEach(to => {
    if (to === pre) return;
    depth[to] = depth[pos] + 1;
    dfs(to, pos);
  });
};

dfs();

for (let i = 0; i < BIT - 1; i++) {
  for (let j = 0; j < N; j++) {
    parents[i + 1][j] = parents[i][parents[i][j]];
  }
}

const lca = (u, v) => {
  if (depth[u] < depth[v]) [u, v] = [v, u];
  const diff = depth[u] - depth[v];
  for (let i = 0; i < BIT; i++) {
    if ((diff >> i) & 1) u = parents[i][u];
  }
  if (u === v) return u;
  for (let i = BIT - 1; i >= 0; i--) {
    if (parents[i][u] == parents[i][v]) continue;
    u = parents[i][u];
    v = parents[i][v];
  }
  return parents[0][u];
}

const Q = Number(input[n++]);
const ans = Array(Q).fill(0);
for (let i = 0; i < Q; i++) {
  const [k, ...v] = input[n++].split(" ").map(Number);
  const w = v.map(j => j - 1).sort((a, b) => orders[a] - orders[b]);
  for (let j = 0; j < k; j++) {
    const parent = lca(w[j], w[(j + 1) % k]);
    ans[i] += depth[w[(j + 1) % k]] - depth[parent];
  }
}

console.log(ans.join("\n"));
